GPII-4130: Add "smoke tests" for preferences server write endpoints.

the-t-in-rtf opened the pull request:
This pull updates the Locust "smoke tests" to evaluate the performance of a wider range of preference server REST endpoints.  This is a prerequisite for evaluating and promoting the work that adds validation to the preference server's endpoints.

See [GPII-4130](https://issues.gpii.net/browse/GPII-4130) for details.

the-t-in-rtf commented:
@stepanstipl, I was imagining that the "write" tests would mainly be run in dev/staging, precisely because they do leave data lying around.  I can see that they would be valuable as part of the pipeline, though.  I'll look into adding cleanup steps, if I can do that without too much trouble, I will.

natarajaya commented:
Thanks @the-t-in-rtf for this attempt to improve our smoke tests!

This looks like a good start, but I share Stepan's concern regarding inclusion of new tests into the CI pipeline. These new tests would have much higher value in case they executed on regular basis. And we need to find a way to clean up after tests if we want Locust to simulate real user behavior (GPII-4158).

natarajaya commented on shared/rakefiles/ci.rake:
```diff
@@ -53,7 +53,7 @@ task :destroy_hard_and_deploy_ci => [:set_vars_ci] do
   begin
     Rake::Task["deploy"].invoke
-    Rake::Task["test_preferences"].invoke
+    Rake::Task["test_preferences_read"].invoke
```
I think it is safe to include `test_preferences_write` here – this task is being executer only against ephemeral environments, so lack for post-test cleanup should not be a problem.

the-t-in-rtf replied:
Great, done.

natarajaya commented on shared/rakefiles/test.rake:
```diff
@@ -11,11 +11,27 @@ task :test do
+desc "[TEST] Run Locust swarm against Preferences service (WRITE) in current cluster"
+task :test_preferences_write => [:set_vars, :check_destroy_allowed] do
+  ENV["TF_VAR_locust_target_host"] = "http://preferences.gpii.svc.cluster.local"
+  ENV["TF_VAR_locust_target_app"] = "preferences"
```
In case this variable stays the same for both tests – there would be no way for us to differentiate results in Stackdriver. Check https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/locust/swarm.tf#L98.

the-t-in-rtf replied:
@natarajaya, once I realised that the write tests were being saved to the same place, I changed the label for the "write" tests.  I left the label for the "read" tests alone, but can easily change that to "preferences_read" if you like.

natarajaya replied:
@the-t-in-rtf I am okay with using new env variable, but the code I am referring to in my previous comment, still uses `locust_target_app` as metric name for [Stackdriver client](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/locust/client.rb#L40-L55).

the-t-in-rtf replied:
So, can I just eliminate the new variable and change `locust_target_app` for the write tests?

the-t-in-rtf replied:
Turns out I can, and did.

the-t-in-rtf commented:
So, I spent a lot more time verifying the write tests locally, and it seems like a) write performance is 50% slower even without validation, and b) we consistently have 4-5 failures for PUT and POST requests in a full run.  As we have never tested writes before, this may unfortunately be normal.  We should discuss more widely and agree how to address this, especially the failure rate.

the-t-in-rtf commented:
@stepanstipl, I just verified the consistent failures late yesterday.  After discussing with Antranig, my plan is to build a container image with my memory usage improvements to see if that improves the performance and/or stability.

cc: @amb26

the-t-in-rtf commented:
@stepanstipl, I worked on this further and think we're in a relatively good spot.  I added checks to confirm that the payload being added via PUT/POST was the same as what we retrieved on a subsequent GET.  In doing so, I found an encoding problem that was corrupting the payloads, and fixed that.  I then found two issues that seem to be timing related.

https://issues.gpii.net/browse/GPII-4200
https://issues.gpii.net/browse/GPII-4201

As a short term workaround for GPII-4200 , I added a hard pause of 375ms between a "write" and the subsequent GET.  This also seems to avoid the error mentioned in GPII-4201.  Antranig and I and others will discuss the performance implications and how we might address them.  For now, though, the updated smoke tests should be ready for further review.

the-t-in-rtf commented on shared/rakefiles/test.rake:
```diff
@@ -30,8 +30,9 @@ task :test_preferences_write => [:set_vars, :check_destroy_allowed] do
-  ENV["TF_VAR_locust_desired_max_response_time"] = "2500"
+  ENV["TF_VAR_locust_desired_total_rps"] = "5"
+  ENV["TF_VAR_locust_desired_median_response_time"] = "1000"
+  ENV["TF_VAR_locust_desired_max_response_time"] = "5000"
```
The overall goal was simply to be able to distinguish "worse than now" performance when testing actual changes, which seems fair for something we've never measured before. 

I suspect that the RPS may be skewed by the hard pauses (`time.sleep` calls) I had to add, if we address whatever is causing that, we probably can safely raise that to 15, which I was able to consistently hit before adding the `time.sleep` calls.

Reviewing the various runs I've conducted and adjusting the levels further is a good idea, especially with data from staging.  I will conduct a few more runs and put together some summary information and links to get us started.

I also think that if we choose to live with a threshold that's too high for now, we should document any observed slowness and work to bring the expected performance up over time.  As an example, 5000ms certainly isn't reasonable, even if it's what we seem to be able to manage at the moment.

the-t-in-rtf commented on gcp/modules/locust/tasks/preferences_write.py:
```diff
@@ -11,50 +13,98 @@ def generate_random_username():
+            else:
+                settings_put_response.failure("The PUT response should contain the original payload we submitted.")
+        else:
+            settings_put_response.failure("A valid prefs payload should have been PUT (error follows):\n" + settings_put_response.text)
```
In the case of the POST response, we need to retrieve the `gpiiKey`  from a successful request, but the rest can be simplified as you suggest.

the-t-in-rtf replied:
NM, assuming that the first failed request breaks the chain, I can just move the `gpiiKey` read out of the `with` block and  use the pattern you suggested.

the-t-in-rtf replied:
OK, done.

the-t-in-rtf commented on gcp/modules/locust/tasks/preferences_write.py:
```diff
@@ -11,50 +13,98 @@ def generate_random_username():
     # We need to catch the results ourselves to pick up the system-generated gpiiKey.
-    with l.client.post(prefs_unkeyed_url, name="3. POST valid prefs without a GPII key.", data=prefs_payload, verify=False, catch_response=True) as valid_post_response:
+    with l.client.post(prefs_unkeyed_url, name="3. POST valid prefs without a GPII key.", json=prefs_in_context, verify=False, catch_response=True) as valid_post_response:
+        print ("POST response:" + json.dumps(valid_post_response.json()) + "...")
```
Those were mainly for early troubleshooting purposes and are now gone.

the-t-in-rtf commented on gcp/modules/locust/tasks/preferences_write.py:
```diff
@@ -11,50 +13,98 @@ def generate_random_username():
     prefs_random_keyed_url = prefs_keyed_url_template.format(random_username)
-    settings_put_response = l.client.put(prefs_random_keyed_url, name="1. PUT valid prefs with a GPII key.", data=prefs_payload, verify=False)
+    with l.client.put(prefs_random_keyed_url, name="1. PUT valid prefs with a GPII key.", json=prefs_in_context, verify=False, catch_response=True) as settings_put_response:
+        print ("PUT response:" + json.dumps(settings_put_response.json()) + "...")
```
Yes, done, throughout.

the-t-in-rtf commented:
OK,  bit of a stats dump, here's the response times for ten successive runs:

```
Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",217,810,930,1000,1100,1300,1600,1900,2100,2200
 "GET 2. GET the prefs we just PUT.",213,330,420,470,500,620,720,810,860,1000
 "POST 3. POST valid prefs without a GPII key.",208,560,640,710,740,850,950,1000,1100,1400
 "GET 4. GET the prefs we just POSTed.",206,300,380,440,510,590,700,800,840,990
 "Total",844,460,600,710,790,960,1100,1400,1600,2200
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",230,810,990,1100,1200,1400,1600,1800,2100,2300
 "GET 2. GET the prefs we just PUT.",228,320,410,440,520,650,700,790,850,920
 "POST 3. POST valid prefs without a GPII key.",223,580,690,730,780,920,990,1300,1300,1500
 "GET 4. GET the prefs we just POSTed.",218,280,350,420,510,630,730,900,950,980
 "Total",899,470,650,730,780,990,1300,1500,1700,2300
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",218,830,1000,1200,1200,1500,1800,2000,2400,2700
 "GET 2. GET the prefs we just PUT.",213,340,430,510,530,630,700,790,850,1300
 "POST 3. POST valid prefs without a GPII key.",213,590,700,760,800,900,1100,1100,1200,1600
 "GET 4. GET the prefs we just POSTed.",212,320,420,480,520,610,650,810,940,1100
 "Total",856,500,620,720,790,1000,1200,1600,1800,2700
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",226,920,1100,1200,1300,1600,1800,1900,2100,3000
 "GET 2. GET the prefs we just PUT.",225,400,490,540,590,710,810,860,960,1300
 "POST 3. POST valid prefs without a GPII key.",223,630,700,750,800,950,1000,1300,1400,1800
 "GET 4. GET the prefs we just POSTed.",220,350,440,520,540,660,710,870,910,1000
 "Total",894,520,660,740,840,1100,1300,1700,1900,3000
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",219,920,1100,1100,1200,1400,1700,1800,2000,2500
 "GET 2. GET the prefs we just PUT.",217,330,420,500,540,700,770,850,940,1000
 "POST 3. POST valid prefs without a GPII key.",213,620,690,730,770,860,910,1100,1100,1300
 "GET 4. GET the prefs we just POSTed.",208,320,380,430,480,610,710,830,870,920
 "Total",857,490,660,750,810,1000,1200,1500,1800,2500
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",225,880,1000,1100,1200,1500,1700,2200,2300,2400
 "GET 2. GET the prefs we just PUT.",220,340,440,530,610,720,840,990,1000,1200
 "POST 3. POST valid prefs without a GPII key.",217,580,680,760,810,920,1000,1100,1200,1400
 "GET 4. GET the prefs we just POSTed.",216,310,390,490,560,680,800,1000,1100,1200
 "Total",878,510,660,760,840,1100,1300,1700,1900,2400
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",154,800,1000,1200,1300,1600,1800,1900,2100,2500
 "GET 2. GET the prefs we just PUT.",145,400,520,610,670,750,790,910,910,1000
 "POST 3. POST valid prefs without a GPII key.",140,630,740,810,850,970,1000,1100,1200,1200
 "GET 4. GET the prefs we just POSTed.",132,390,510,580,610,720,810,900,910,1100
 "Total",571,540,690,770,820,1000,1400,1700,1800,2500
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",222,840,950,1100,1200,1500,1700,1800,1900,2500
 "GET 2. GET the prefs we just PUT.",220,320,400,480,540,700,810,1000,1200,1200
 "POST 3. POST valid prefs without a GPII key.",219,590,680,730,780,930,1100,1200,1200,1500
 "GET 4. GET the prefs we just POSTed.",214,300,390,450,500,670,790,820,870,960
 "Total",875,490,630,740,810,990,1200,1600,1800,2500
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",228,900,1000,1100,1200,1500,1800,1900,2700,2900
 "GET 2. GET the prefs we just PUT.",226,320,420,480,540,700,760,940,980,1200
 "POST 3. POST valid prefs without a GPII key.",225,590,690,740,780,880,940,1000,1100,1300
 "GET 4. GET the prefs we just POSTed.",222,310,390,440,490,620,690,800,910,1300
 "Total",901,500,660,740,820,1000,1200,1600,1800,2900
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",208,920,1100,1200,1400,2200,2600,2900,3000,3100
 "GET 2. GET the prefs we just PUT.",206,320,420,500,550,730,930,1100,1100,1600
 "POST 3. POST valid prefs without a GPII key.",204,610,730,800,900,1100,1300,1500,2000,2100
 "GET 4. GET the prefs we just POSTed.",200,290,360,410,470,710,830,960,1100,1300
 "Total",818,500,700,810,900,1100,1600,2400,2700,3100
--
 Stats distribution:
 "Name","# requests","50%","66%","75%","80%","90%","95%","98%","99%","100%"
 "PUT 1. PUT valid prefs with a GPII key.",216,940,1200,1300,1300,1600,1800,2400,2500,2800
 "GET 2. GET the prefs we just PUT.",214,340,470,530,600,780,880,940,1000,1100
 "POST 3. POST valid prefs without a GPII key.",211,610,720,800,840,920,1000,1100,1300,1700
 "GET 4. GET the prefs we just POSTed.",211,330,430,530,540,700,830,920,1000,1100
 "Total",852,520,710,810,880,1100,1300,1700,1900,2800
```
Here are just the totals:

```
 "Total",844,460,600,710,790,960,1100,1400,1600,2200
 "Total",899,470,650,730,780,990,1300,1500,1700,2300
 "Total",856,500,620,720,790,1000,1200,1600,1800,2700
 "Total",894,520,660,740,840,1100,1300,1700,1900,3000
 "Total",857,490,660,750,810,1000,1200,1500,1800,2500
 "Total",878,510,660,760,840,1100,1300,1700,1900,2400
 "Total",571,540,690,770,820,1000,1400,1700,1800,2500
 "Total",875,490,630,740,810,990,1200,1600,1800,2500
 "Total",901,500,660,740,820,1000,1200,1600,1800,2900
 "Total",818,500,700,810,900,1100,1600,2400,2700,3100
 "Total",852,520,710,810,880,1100,1300,1700,1900,2800
```
Maybe 3500ms is a good threshold for now?  And here's the `total_rps` for those runs:

```
   "total_rps": 15.0,
   "total_rps": 20.3,
   "total_rps": 20.0,
   "total_rps": 18.7,
   "total_rps": 18.6,
   "total_rps": 20.5,
   "total_rps": 0.0,
   "total_rps": 19.8,
   "total_rps": 21.0,
   "total_rps": 21.6,
   "total_rps": 20.4,
```

The run with 0 RPS failed outright, the rest seem to easily beat 15.  I'm happy with using that depending on what we see in staging.

the-t-in-rtf commented:
Thanks, @stepanstipl.  I see two variations on errors.  First:

```
    {
      "error": "\"HTTPError('404 Client Error: Not Found for url: http://preferences.gpii.svc.cluster.local/preferences/extra5741',)\"",
      "method": "GET",
      "name": "2. GET the prefs we just PUT.",
      "occurences": 1
    }
```

This seems to me to be precisely the "read after write" error that my time.sleep calls are meant to avoid.  Although we will eventually need to address the underlying issue, for now I can adjust the `time.sleep` wait upwards to see if that helps.

Here's the second error:

```
    {
      "error": "\"HTTPError('503 Server Error: Service Unavailable for url: http://preferences.gpii.svc.cluster.local/preferences/extra8574',)\"",
      "method": "PUT",
      "name": "1. PUT valid prefs with a GPII key.",
      "occurences": 1
    },
```

I have also seen this, and I suspect it may be related to the memory issues [my fix](https://github.com/GPII/universal/pull/811) for [GPII-4062](https://issues.gpii.net/browse/GPII-4062) addresses.  I can test with an updated universal image and confirm whether that helps in repeated runs.

the-t-in-rtf commented:
My update to the pause between POST/PUT and the next GET seems to have eliminated the 404 errors in 5 back to back runs.

However, even with the timing change and with my memory fixes, I still see the 503s, I'll look through the instance's logs to see if I can figure anything out:

```
 {
       "error": "\"HTTPError('503 Server Error: Service Unavailable for url: http://preferences.gpii.svc.cluster.local/preferences/chancey3836',)\"",
       "method": "PUT",
       "name": "1. PUT valid prefs with a GPII key.",
       "occurences": 1
}
```

the-t-in-rtf commented:
In reviewing the logs in GCP, it really seems like the 503 errors are the "error marking thread to request" problems I wrote up as [GPII-4201](https://issues.gpii.net/browse/GPII-4201).  @amb26, did you have any suggestions for narrowing that down?

the-t-in-rtf commented:
Based on discussions with @stepanstipl and others, I adjusted the expected performance for the new preferences "write" tests to match a level that does not result in intermittent failures when running with the current universal docker container.  This should be ready for further review.

the-t-in-rtf commented:
@stepanstipl (and anyone else interested in reviewing), as I mentioned on Slack yesterday, I think I've pinned down the number of simulated users and expected performance.  It would be nice to get this merged and start talking about evaluating the backlog of actual changes to universal.

stepanstipl commented:
thanks @the-t-in-rtf, we'll need to run some tests in `stg` to see if those failures are gone. I'll post results here once done.

natarajaya commented:
LGTM!
